File: concurrency/javascript/blocking.js
class Blocking {

    start() {
        const startTime = new Date();

        while (true) {
            console.log(`Rendring UI on... ${process.pid}`);
            const now = new Date();

            if (now - startTime >= 5000 && now - startTime <= 6000) {
                this.doRequest();
            }

            this.sleep(1000);
        }
    }

    doRequest() {
        console.log(`Initiated the request.... on  ${process.pid}`);
        this.sleep(3000);
        console.log(`Finished the request.... on ${process.pid}`);
    }

    sleep(ms) {
        var start = new Date().getTime(), expire = start + ms;
        while (new Date().getTime() < expire) { }
        return;
    }
}

export { Blocking };

File: concurrency/c/blocking.c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <unistd.h>

long long get_current_time_ms() {
  struct timeval tv;
  gettimeofday(&tv, NULL);
  return (tv.tv_sec * 1000) + (tv.tv_usec / 1000);
}

void doRequest() {
  printf("Initiated the request.... On %ld\n", pthread_self());
  sleep(3);
  printf("Finished the request On %ld\n", pthread_self());
}

void start() {
  long long start = get_current_time_ms();

  while (1) {
    printf("Rendering UI On... %ld\n", pthread_self());
    long long now = get_current_time_ms();

    if (now - start >= 5000 && now - start < 6000) {
      doRequest();
    }

    sleep(1);
  }
}

int main(int argc, char *argv[]) {
  start();
  return EXIT_SUCCESS;
}
